tools/build.py

```diff
@@ -0,0 +1,645 @@
+
+    log.info("Installing emsdk...")
+    _run_subprocess([emsdk_file, "install", args.emsdk_version], cwd=args.emsdk_path)
+    log.info("Activating emsdk...")
+    _run_subprocess([emsdk_file, "activate", args.emsdk_version], cwd=args.emsdk_path)
+
+
+def _generate_build_tree(cmake_path: Path,
+                         source_dir: Path,
+                         build_dir: Path,
+                         configs: Set[str],
+                         cmake_extra_defines: List[str],
+                         args,
+                         cmake_extra_args: List[str]
+                         ):
+    log.info("Generating CMake build tree")
+
+    cmake_args = [
+        str(cmake_path),
+        str(source_dir),
+        # Define Python_EXECUTABLE so find_package(python3 ...) will use the same version of python being used to
+        # run this script
+        "-DPython_EXECUTABLE=" + sys.executable,
+        "-DOCOS_ENABLE_SELECTED_OPLIST=" + ("ON" if _is_reduced_ops_build(args) else "OFF"),
+    ]
+
+    if args.enable_cxx_tests:
+        cmake_args.append("-DOCOS_ENABLE_CTEST=ON")
+        ort_lib_dir = _validate_cxx_test_args(args)
+        if ort_lib_dir:
+            cmake_args.append(f"-DONNXRUNTIME_LIB_DIR={str(ort_lib_dir)}")
+
+        if args.onnxruntime_version:
+            cmake_args.append(f"-DONNXRUNTIME_VER={args.onnxruntime_version}")
+
+    if args.android:
+        if not args.android_ndk_path:
+            raise UsageError("android_ndk_path is required to build for Android")
+        if not args.android_home:
+            raise UsageError("android_home is required to build for Android")
+
+        android_home = args.android_home.resolve(strict=True)
+        android_ndk_path = args.android_ndk_path.resolve(strict=True)
+
+        if not android_home.is_dir() or not android_ndk_path.is_dir():
+            raise UsageError("Android home and NDK paths must be directories.")
+
+        ndk_version = android_ndk_path.name  # NDK version is inferred from the folder name
+
+        cmake_args += [
+            "-DOCOS_BUILD_ANDROID=ON",
+            "-DCMAKE_TOOLCHAIN_FILE="
+            + str((args.android_ndk_path / "build" / "cmake" / "android.toolchain.cmake").resolve(strict=True)),
+            "-DANDROID_NDK_VERSION=" + str(ndk_version),
+            "-DANDROID_PLATFORM=android-" + str(args.android_api),
+            "-DANDROID_ABI=" + str(args.android_abi)
+        ]
+
+    if is_macOS():
+        cmake_args.append("-DOCOS_BUILD_APPLE_FRAMEWORK=" + ("ON" if args.build_apple_framework else "OFF"))
+
+    if args.ios:
+        required_args = [
+            args.ios_sysroot,
+            args.apple_deploy_target,
+        ]
+
+        arg_names = [
+            "--ios_sysroot          " + "<the location or name of the macOS platform SDK>",
+            "--apple_deploy_target  " + "<the minimum version of the target platform>",
```

Comment:
I think we'd also need the target arch too. can add it later when enabling iOS builds here